docs/rfcs/0004-agent-flow.md

```diff
@@ -0,0 +1,162 @@
+
+## Overarching Problem Statement
+
+The Agents configuration and onboarding is difficult to use. Viewing the effect of configuration changes on telemetry data is difficult. Making the configuration simplier, composable and intuitive allieviates these concerns.
+
+
+## Description
+
+Agent Flow is intended to solve real world needs that the Grafana Agent team have idenfified in conversations with users and developers. 
+
+These broadly include 
+
+- Lack of introspection within the agent
+    - Questions about what telemetry data are being sent
+    - Are rules applying correctly?
+    - How does filtering work?
+- Users have been requesting additional capabilites and adding new features is hard due to coupling between systems, some examples include
+    - Remote Write, Different Input Formats
+    - Different output formats
+    - Filtering ala Relabel Configs is complex and hard to figure out when they occur
+- Lack of understanding how telemetry data moves through agent
+    - Other systems use pipeline/extensions to allow users to understand how data moves through the system
+
+# 1. Introduction and Goals 
+
+This design document describes Agent Flow, which is a system for describing the execution and configuration of telemetry data, along with any secondary configuration required to configure the Grafana Agent. 
+
+Agent Flow refers to both the execution, configuration and visual configurator of data flow.
+
+The primary goals of Agent Flow are
+
+1. Allow users to more easily understand the impact of their configuration
+2. Allow users to collect metrics from all running integrations across a  set of agents without installing extra dependencies
+3. Allow users to run components based on a dynamic environment
+4. Allow developers to easily add components to the system
+5. Ensure the Agent still maintains high performance on all platforms that the Agent currently supports
+6. Flow configs must be easily machine-readable and machine-writable to support tooling, such as formatting and GUI
+
+Goals of this design document
+
+* Define high level concepts and goals
+* Define high level concepts of the execution path
+
+This document represents ideals and not technical implementation. 
+
+Non Goals of this design document
+
+* Define a technical implementation or configuration implementation
+
+# 2. Broad Solution Path
+
+Conversation around should the components be assembled via message passing, via expressions, or a hybrid approach.
+
+**Note: Consider all examples pseudoconfig**
+
+## 2.1 Expression Based
+
+Expression based is writing expressions that allow referencing other components streams/outputs/values and using them directly. 
+
+**Pros**
+
+* Easier to Implement, evaluating expressions can map directly to existing config structs
+* Components are more reusable, you can pass basic types arounds (string, int, bool)
+
+**Cons**
+* Harder for users to wire things together
+* Harder to build a GUI for
+
+
+## 2.2 Message Based
```

Comment:
One thing that just came to mind: it's less obvious to users what fields a connection between two message-based components is overriding; you'd have to pull up the docs to figure out what the connection sets. Comparatively, in expressions, the settings being made dynamic is explicit and can be inferred just by reading the config file.

Reply:
Ideally, for most components, the user shouldn't need to worry about it. I do think there will be cases where they do and that will not be ideal.

Reply:
The most common case I see is a user configuring a component: they need to make sure they don't configure both the static field and the connection, so they'll need to pull up the documentation to see what values are still required.

I think this will be the case when configuring a new component or editing an existing component to have a connection or to remove a connection and make it static.